libc/src/__support/FPUtil/generic/div.h

```diff
@@ -0,0 +1,180 @@
+//===-- Division of IEEE 754 floating-point numbers -------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef LLVM_LIBC_SRC___SUPPORT_FPUTIL_GENERIC_DIV_H
+#define LLVM_LIBC_SRC___SUPPORT_FPUTIL_GENERIC_DIV_H
+
+#include "hdr/fenv_macros.h"
+#include "src/__support/CPP/bit.h"
+#include "src/__support/CPP/type_traits.h"
+#include "src/__support/FPUtil/FEnvImpl.h"
+#include "src/__support/FPUtil/FPBits.h"
+#include "src/__support/FPUtil/dyadic_float.h"
+#include "src/__support/FPUtil/rounding_mode.h"
+
+namespace LIBC_NAMESPACE::fputil::generic {
+
+template <typename OutType, typename InType>
+cpp::enable_if_t<cpp::is_floating_point_v<OutType> &&
+                     cpp::is_floating_point_v<InType> &&
+                     sizeof(OutType) <= sizeof(InType),
+                 OutType>
+div(InType x, InType y) {
+  using OutFPBits = FPBits<OutType>;
+  using OutStorageType = typename OutFPBits::StorageType;
+  using InFPBits = FPBits<InType>;
+  using InStorageType = typename InFPBits::StorageType;
+  using DyadicFloat =
+      DyadicFloat<cpp::bit_ceil(static_cast<size_t>(InFPBits::FRACTION_LEN))>;
+  using DyadicMantissaType = typename DyadicFloat::MantissaType;
+
+  // +1 for the implicit bit.
+  constexpr int DYADIC_EXTRA_MANTISSA_LEN =
+      DyadicMantissaType::BITS - (InFPBits::FRACTION_LEN + 1);
+  // +1 for the extra fractional bit in q.
+  constexpr int Q_EXTRA_FRACTION_LEN =
+      InFPBits::FRACTION_LEN + 1 - OutFPBits::FRACTION_LEN;
+
+  InFPBits x_bits(x);
+  InFPBits y_bits(y);
+
+  if (x_bits.is_nan() || y_bits.is_nan()) {
+    if (x_bits.is_signaling_nan() || y_bits.is_signaling_nan())
```

Comment:
I should try putting all of these `if`s for special values under a single `if (LIBC_UNLIKELY(...))` and running a benchmark to compare.

Reply:
+1

Reply:
I pushed the change to `if (LIBC_UNLIKELY(...))`. It is indeed faster:

```
 Performance tests with inputs in denormal range:
-- My function --
     Total time      : 590356165 ns 
     Average runtime : 29.5178 ns/op 
     Ops per second  : 33877887 op/s 
-- Other function --
     Total time      : 612659035 ns 
     Average runtime : 30.6329 ns/op 
     Ops per second  : 32644617 op/s 
-- Average runtime ratio --
     Mine / Other's  : 0.963597 

 Performance tests with inputs in normal range:
-- My function --
     Total time      : 145477558 ns 
     Average runtime : 7.27387 ns/op 
     Ops per second  : 137478386 op/s 
-- Other function --
     Total time      : 172587200 ns 
     Average runtime : 8.62935 ns/op 
     Ops per second  : 115883564 op/s 
-- Average runtime ratio --
     Mine / Other's  : 0.842922 

 Performance tests with inputs in normal range with exponents close to each other:
-- My function --
     Total time      : 522964890 ns 
     Average runtime : 26.1482 ns/op 
     Ops per second  : 38243523 op/s 
-- Other function --
     Total time      : 550152260 ns 
     Average runtime : 27.5076 ns/op 
     Ops per second  : 36353608 op/s 
-- Average runtime ratio --
     Mine / Other's  : 0.950582 
```

Reply:
It's nice to have perf tests to confirm improvements.

Reply:
Even with the changes to `DyadicFloat::operator T()`, the new implementation with fewer iterations is still faster than the previous one:

```
 Performance tests with inputs in denormal range:
-- My function --
     Total time      : 232472260 ns 
     Average runtime : 46.4944 ns/op 
     Ops per second  : 21507964 op/s 
-- Other function --
     Total time      : 519384676 ns 
     Average runtime : 103.877 ns/op 
     Ops per second  : 9626785 op/s 
-- Average runtime ratio --
     Mine / Other's  : 0.447592 

 Performance tests with inputs in normal range:
-- My function --
     Total time      : 511714277 ns 
     Average runtime : 102.343 ns/op 
     Ops per second  : 9771087 op/s 
-- Other function --
     Total time      : 637747828 ns 
     Average runtime : 127.549 ns/op 
     Ops per second  : 7840097 op/s 
-- Average runtime ratio --
     Mine / Other's  : 0.802377 

 Performance tests with inputs in normal range with exponents close to each other:
-- My function --
     Total time      : 309770755 ns 
     Average runtime : 61.9541 ns/op 
     Ops per second  : 16140984 op/s 
-- Other function --
     Total time      : 512884190 ns 
     Average runtime : 102.577 ns/op 
     Ops per second  : 9748799 op/s 
-- Average runtime ratio --
     Mine / Other's  : 0.603978 
```

Reply:
Performance test of the current version (https://github.com/llvm/llvm-project/pull/96131/commits/6439428c4007ff8df13a5c31cb9c37e5ecdb5e07) vs https://github.com/llvm/llvm-project/pull/96131/commits/4ff508d039ac82b9489ece2337231cff259deb8f:

```
 Performance tests with inputs in denormal range:
-- My function --
     Total time      : 224276882 ns 
     Average runtime : 44.8553 ns/op 
     Ops per second  : 22293893 op/s 
-- Other function --
     Total time      : 520018166 ns 
     Average runtime : 104.004 ns/op 
     Ops per second  : 9615058 op/s 
-- Average runtime ratio --
     Mine / Other's  : 0.431287 

 Performance tests with inputs in normal range:
-- My function --
     Total time      : 506187486 ns 
     Average runtime : 101.237 ns/op 
     Ops per second  : 9877772 op/s 
-- Other function --
     Total time      : 642170120 ns 
     Average runtime : 128.434 ns/op 
     Ops per second  : 7786106 op/s 
-- Average runtime ratio --
     Mine / Other's  : 0.788245 

 Performance tests with inputs in normal range with exponents close to each other:
-- My function --
     Total time      : 303552644 ns 
     Average runtime : 60.7105 ns/op 
     Ops per second  : 16471623 op/s 
-- Other function --
     Total time      : 513769699 ns 
     Average runtime : 102.754 ns/op 
     Ops per second  : 9731996 op/s 
-- Average runtime ratio --
     Mine / Other's  : 0.590834 
```